api/v1alpha1/stage_types.go

```diff
@@ -155,15 +155,15 @@ type ImageSubscription struct {
 	// value in this field MUST NOT include an image tag. This field is required.
 	//
 	//+kubebuilder:validation:MinLength=1
-	//+kubebuilder:validation:Pattern=`^(([\w\d\.-]+)(:[\d]+)?/)?[a-z0-9-]+(/[a-z0-9-]+)*$`
+	//+kubebuilder:validation:Pattern=`^(([\w\d\.-]+)(:[\d]+)?/)?[a-z0-9-_\.]+(/[a-z0-9-_\.]+)*$`
```

Comment:
Yours is a clear improvement, but I think we can tighten this up further.

I propose: `^(\w+([\.-]\w+)*(:[\d]+)?/)?(\w+([\.-]\w+)*)(/\w+([\.-]\w+)*)*$`

Broken down:

| Expression Part | Explanation |
|-----------------|-------------|
| `^` | |
| `(\w+([\.-]\w+)*(:[\d]+)?/)?` | Optional host with optional port number  |
| `(\w+([\.-]\w+)*)` | Required token starts and ends with a word character; this could be image name, or could be "owner" if followed by more stuff |
| `(/\w+([\.-]\w+)*)*` | Any number of additional tokens separated by forward slash; each token must begin and end with a word character |
| `$` | |

![Screenshot 2023-10-04 at 11 54 58 AM](https://github.com/akuity/kargo/assets/1821014/0bd312d1-488c-404e-8cb7-bc6720fb7f57)

What do you think about this? Although tighter overall, it does relax the things you were looking for, I think?

Reply:
I tried to get fancy squashing inside my branch so your comments got disassociated - but I accepted all your suggestions.